auth/jwt.go

```diff
@@ -50,4 +77,123 @@ type LocalJWTAuthProvider struct {
+	}
+
+	var key *jose.JSONWebKey
+	if len(l.Keys) == 1 {
+		key = &l.Keys[0]
+	} else {
+		for i, test := range l.Keys {
+			if test.KeyID == kid {
+				key = &l.Keys[i]
+				break
+			}
+		}
+	}
+	if key == nil {
+		base.DebugfCtx(ctx, base.KeyAuth, "Rejecting JWT - no matching key")
+		return nil, fmt.Errorf("no matching key")
+	}
+	if key.Algorithm != alg {
+		base.DebugfCtx(ctx, base.KeyAuth, "Rejecting JWT - key alg mismatch (expected %v got %v)", base.UD(alg), base.UD(key.Algorithm))
+		return nil, fmt.Errorf("key alg mismatch (expected %s got %s)", alg, key.Algorithm)
+	}
+	if key.Use != "" && key.Use != "sig" {
+		base.DebugfCtx(ctx, base.KeyAuth, "Rejecting JWT - invalid key 'use' (%q)", base.UD(key.Use))
+		return nil, fmt.Errorf("invalid key use")
+	}
+
+	rawPayload, err := jws.Verify(key.Key)
+	if err != nil {
+		base.DebugfCtx(ctx, base.KeyAuth, "Rejecting JWT - invalid signature: %v", err)
+		return nil, fmt.Errorf("invalid JWT signature: %w", err)
+	}
+
+	identityJSON, err := UnmarshalIdentityJSON(rawPayload)
+	if err != nil {
+		base.DebugfCtx(ctx, base.KeyAuth, "Rejecting JWT - failed to parse payload: %v", err)
+		return nil, fmt.Errorf("invalid JWT payload: %w", err)
+	}
+
+	expected := jwt.Expected{
+		Issuer: l.Issuer,
+		Time:   time.Now(),
+	}
+	if *l.ClientID != "" {
+		expected.Audience = jwt.Audience{*l.ClientID}
+	}
+	if err := identityJSON.ToClaims().Validate(expected); err != nil {
+		base.DebugfCtx(ctx, base.KeyAuth, "Rejecting JWT - invalid: %v", err)
+		return nil, err
+	}
+	base.DebugfCtx(ctx, base.KeyAuth, "JWT successfully parsed and validated (issuer %q, subject %q)", base.UD(identityJSON.Issuer), base.UD(identityJSON.Subject))
+
+	return &Identity{
+		Issuer:   identityJSON.Issuer,
+		Audience: identityJSON.Audience,
+		Subject:  identityJSON.Subject,
+		Expiry:   time.Time(identityJSON.Expiry),
+		IssuedAt: time.Time(identityJSON.IssuedAt),
+		Email:    identityJSON.Email,
+		Claims:   identityJSON.Claims,
+	}, nil
+}
+
+func (l *LocalJWTAuthProvider) common() JWTConfigCommon {
+	return l.JWTConfigCommon
+}
+
+func (l *LocalJWTAuthProvider) InitUserPrefix(ctx context.Context, name string) {
+	if l.UserPrefix != "" || l.UsernameClaim != "" {
+		return
+	}
+
+	issuerURL, err := url.ParseRequestURI(l.Issuer)
+	if err != nil {
+		base.WarnfCtx(ctx, "Unable to parse issuer URI when initializing user prefix - using provider name")
+		l.UserPrefix = name
+		return
+	}
+	l.UserPrefix = issuerURL.Host + issuerURL.Path
+
+	// If the prefix contains forward slash or underscore, it's not valid as-is for a username: forward slash
+	// breaks the REST API, underscore breaks uniqueness of "[prefix]_[sub]".  URL encode the prefix to cover
+	// this scenario
+	l.UserPrefix = url.QueryEscape(l.UserPrefix)
+}
+
 type LocalJWTConfig map[string]*LocalJWTAuthProvider
+
+type jwtAuthenticator interface {
+	verifyToken(ctx context.Context, token string, callbackURLFunc OIDCCallbackURLFunc) (*Identity, error)
```

Comment:
Having something OIDC-specific on the common interface suggests that the interface might not be at exactly the right level.  In this case I think it's to support lazy-loading of the OIDC provider?  I wonder if that could be handled prior to the verifyToken invocation for the OIDC flows.

Reply:
It is a bit of a smell, I agree. The problem is that `buildCallbackURLFunc` is required for `OIDCProvider.GetClient`(which then uses it in `initOIDCClient` and creates `OIDCClient` based on it) and removing that requirement would require extensive changes to `OIDCClient` which seemed beyond the scope of this PR. My gut feeling is that this would be made redundant by CBG-2170 - though I'm happy to reconsider if you feel strongly.